core/utils/index_utils.hpp

```diff
@@ -24,6 +24,206 @@
+    iterator_t last() const noexcept { return segments.second; }
+  
+    void advance() noexcept;
+  
+    size_t mergeBytes { 0 };
+    double skew { 0.0 };
+    size_t delCount { 0 };
+    double mergeScore { 0.0 };
+    double cost { 0.0 };
+    bool initialized { false };
+  
+    range_t segments;
+  };
+  
+  size_t getConsolidationTier(size_t num);
+
+
+  template<typename Type>
+  using Iterator = std::vector<Type>::const_iterator;
+
+  //  A tier is nothing but a pair of (start, end) iterator
+  //  pointing to segments in a sorted array of segments. This pair
+  //  denotes a range of segments that are to be considered as part
+  //  of the same group.
+  template<typename TierElementType>
+  using Tier = std::pair<Iterator<TierElementType>, Iterator<TierElementType>>;
+  
+  //  mapToTiers() goes over a sorted (by size) list of segments
+  //  and groups these segments into different tiers (representing different
+  //  size groups). We use powers of 4 for tiers with the first tier
+  //  being 0-4M followed by 4-16M, 16-64M, 64-256M and so on.
+  //
+  //  The first argument is a vector of segments.
+  //  This function only uses the size of a segment to make the decision
+  //  as to which tier a segment should be a part of. That is what brings
+  //  us to the 2nd argument. GetSegmentSize is a function/lambda that takes
+  //  as argument a segment and returns its size.
+  //  From the perspective of the mapToTiers() function, the segment is an
+  //  opaque object.
+  //
+  //  The return type is a vector<Tier> where Tier which in turn is a
+  //  pair of iterators pointing to a range of segments that are part of
+  //  a tier.
+  //  Eg.
+  //      Segments: 10K 20K 500K 3.5M 200M 250M 255M 300M 310M 500M 1024M
+  //      Tiers:   |<---- 0-4M ----->|<-- 64-256M ->|<---- 256-1024M ---->|
+  //
+  template<typename Segment, typename GetSegmentSize>
+  requires std::invocable<GetSegmentSize, const Segment&>
+  std::vector<Tier<Segment>> mapToTiers(
+    const std::vector<Segment>& sorted_segments,   //  assumed sorted in asc order
+    GetSegmentSize getSize
+    )
+  {
+    if (sorted_segments.empty())
+      return {};
+
+    size_t currentTier;
+    
+    auto start = sorted_segments.begin();
+    currentTier = getConsolidationTier(getSize(*start));
+
+    std::vector<Tier<Segment>> tiers;
+
+    for (auto idx = start; idx != sorted_segments.end();) {
+
+        if (getSize(*idx) <= currentTier) {
+            idx++;
+            continue;
+        }
+
+        tiers.emplace_back(start, idx - 1);
+
+        //  The next tier may not necessarily be in the
+        //  next power of 4.
+        //  Consider this example,
+        //     [2, 4, 6, 8, 900]
+        //  While the 2, 4 fall in the 0-4 tier and 6, 8 fall
+        //  in the 4-16 tier, the last segment falls in
+        //  the [256-1024] tier.
+
+        currentTier = getConsolidationTier(getSize(*idx));
+        start = idx++;
+    }
```

Comment:
As discussed: finding the tier-boundaries could be done by binary search, possibly utilizing `std::lower_bound` / `std::upper_bound`.